### PR TITLE
GIFTI format support for reading data arrays with intent NIFTI_INTENT_NONE

### LINKED ISSUE #2103: Extend GiftiMeshIO to read data with intent NIFTI_INTENT_NONE into PointData
### Description

GiftiMeshIO doesn't read data from a DataArray with Intent="NIFTI_INTENT_NONE" (data written by cat12 Matlab-Toolbox)

### Expected behavior

Gifti format documentation states that all data from DataArrays with Intent starting with "NIFTI_INTENT" is functional data. I suggest to read this data into PointData of the mesh.

### Actual behavior

DataArray with intent "NIFTI_INTENT_NONE" is skipped

### Additional Information

-

### PR DESCRIPTION
**Pull request by crossmanith**
BUG: read data arrays with intent NIFTI_INTENT_NONE for point or cell data

fixes #2103

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx [hunk 1]
```diff
@@ -168,45 +175,11 @@ GiftiMeshIO::ReadMeshInformation()
...
-          break;
-        case NIFTI_TYPE_FLOAT64:
-          this->m_PointComponentType = IOComponentEnum::DOUBLE;
-          break;
-        case NIFTI_TYPE_FLOAT128:
-          this->m_PointComponentType = IOComponentEnum::LDOUBLE;
-          break;
-        default:
-          itkExceptionMacro(<< "Unknown point component type");
+        itkExceptionMacro(<< "Invalid datatype in data array " << ii << " detected.");
```
Comment (dzenanz) [inline Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx:185]: A more descriptive message here might be helpful. Something like: "Pointset ii is expected to contain only scalars".

File: Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx [hunk 2]
```diff
@@ -236,45 +209,10 @@ GiftiMeshIO::ReadMeshInformation()
...
-        case NIFTI_TYPE_FLOAT64:
-          this->m_CellComponentType = IOComponentEnum::DOUBLE;
-          break;
-        case NIFTI_TYPE_FLOAT128:
-          this->m_CellComponentType = IOComponentEnum::LDOUBLE;
-          break;
-        default:
-          gifti_free_image(m_GiftiImage);
-          itkExceptionMacro(<< "Unknown cell component type");
+        itkExceptionMacro(<< "Invalid datatype in data array " << ii << " detected.");
```
Comment (dzenanz) [inline Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx:219]: Preferably more descriptive, see previous suggestion.

File: Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx [hunk 3]
```diff
@@ -140,6 +140,10 @@ GiftiMeshIO::SetLabelNameTable(const LabelNameContainer * labelMap)
   this->Modified();
 }
 
+// TODO:  Need to be able to read/write RGB images into ITK.
```
Comment (dzenanz) [inline Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx:143]: This doesn't look like the right place for this comment.

File: Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx [hunk 4]
```diff
@@ -258,15 +258,15 @@ GiftiMeshIO::ReadMeshInformation()
           this->m_UpdatePointData = true;
           this->m_PointPixelComponentType = GetComponentTypeFromGifti(m_GiftiImage->darray[ii]->datatype);
           this->m_PointPixelType = GetPixelTypeFromGifti(m_GiftiImage->darray[ii]->datatype);
-          this->m_NumberOfPointPixelComponents = GetNumberOfPixelComponentsFromGifti(m_GiftiImage->darray[ii]->datatype);
+          this->m_NumberOfPointPixelComponents =
```
Comment (dzenanz) [inline Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx:261]: All these style commits at the end should be squashed into one commit with `STYLE: ` prefix in commit message.

**Review by dzenanz — COMMENTED** [review]
Good refactoring and update. Some in-line suggestions.

File: Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx (hunk 3 above)
Comment (crossmanith) [reply Modules/IO/MeshGifti/src/itkGiftiMeshIO.cxx:143]: Any objections to just remove this comment?

**Comment by crossmanith** [comment]
What about ITK.Windows failed? Am I expected to fix something?

**Comment by dzenanz** [comment]
There is one [test failure](https://open.cdash.org/test/294066944), which seems unrelated to this PR. You can ignore it.

**Comment by thewtex** [comment]
@crossmanith thank you for the contribution! :pray: 

Could the branch please be rebased to pass the commit message formatting requirements?

See:

https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#revise-a-topic
https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#commit-messages

**Comment by crossmanith** [comment]
Before I was able to create the PR I had to add some commits to be in sync with the coding style - I did this on github directly. Don't know, what went wrong after that, but I started with a fresh clone, applied my changes again and grouped them in 6 instead of 8 commits with well formed commit messages. Now I can't push to branch ITK_2103. Would it be a good or a bad idea to delete branch ITK_2103, recreate it and push to this new branch?

**Comment by dzenanz** [comment]
You should force-push to the existing branch (command line option `--force-with-lease`, or `--force`).

**Review by thewtex — APPROVED** [review]
@crossmanith thank you for the contribution!

**Comment by jhlegarreta** [comment]
Folks, I've seen that the coverage has increased from 89.7% on 89.85% to comparing the reports on [Wednesday, January 06 2021](https://open.cdash.org/viewCoverage.php?buildid=6967258&filtercount=1&field1=filename/string&compare1=63&value1=&status=-1) and [Thursday, January 07 202](https://open.cdash.org/viewCoverage.php?buildid=6969591&filtercount=1&field1=filename/string&compare1=63&value1=&status=-1). There are 248 less lines untested, but only 12 more tested (site: `CDash Server`, build `Aggregate Coverage`; numbers are similar for `blaster.kitware`, build `Code Coverage`; and the increase in perecentage is also similar for `AWS-BATCH`, build `CentOS Coverage`)

The commit hash corresponding to the `AWS-BATCH` [site](https://open.cdash.org/index.php?project=Insight#!#Nightly) points to [414ad0 ](https://github.com/InsightSoftwareConsortium/ITK/commit/414ad0f5b6189e770ed925d5b1026875ecba1902), contained in this PR. (The warnings in the build are unrelated to this PR).

The `Modules/IO/MeshGifti/src` directory has passed from 31.31%, 601/875 to 45.37% | 348/637 (coverage | lines not covered) between those two days. Have not thoroughly checked all directories.

Coverage increases are great news (and such great leaps forward are :100:; not easy to achieve), but at times large changes are indicative of issues, such as files not being compiled, files that have been moved out of the ITK proper source tree, etc. I have often merged (coverage improvement) PRs because tests were passing in GitHub, but have found later in the `CDash`dashboard that some non-default modules were not compiling, seeing such high variations in the coverage as a consequence.

Looks like it is not the case in here: I see in the PR that the refactoring has allowed to better re-use the code, which may help to explain the less lines existing in the module.

However, I just want to make sure that the increase in coverage is expected or a nice benefit of this PR. Thanks for the tremendous effort in this PR.

**Comment by dzenanz** [comment]
Refactoring which significantly reduces lines of code is rare. I have only seen it in IO modules, which were created by a variety of people.

Simplification refactoring's main benefit is reducing possibility of copy-paste errors. Increased coverage is just one indication of the improvement.

**Comment by jhlegarreta** [comment]
Sounds good to me :+1:. These efforts certainly deserve a :medal_sports: !

**Pull request merged**